### PR TITLE
Laboratory work #1, Margarita Velmisova - 22FPL1

### PR DESCRIPTION
**Pull request by velgarita**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -11,6 +13,14 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:16]: Let's try to use early return: https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -11,6 +13,14 @@ def tokenize(text: str) -> list[str] | None:
...
+        tokens = []
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:17]: Can you try to refactor the code using list comprehention?

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -19,6 +29,17 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list):
+        return None
+
+    if not all(isinstance(token, str) for token in tokens):
+        return None
+
+    frequency = {}
+    for token in tokens:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:39]: Again, try dict comprehension

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -28,6 +49,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     :param text: a text
     :return: a dictionary with two keys – name, freq
     """
+    if isinstance(language, str) and isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:52]: please, use early return

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -37,32 +63,75 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     :param actual: a list of actual values
     :return: the score
     """
+    if isinstance(predicted, list) and isinstance(actual, list) and len(predicted) == len(actual):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:66]: again - early return

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -37,32 +63,75 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+        sum_difference = 0
+        for i, value in enumerate(predicted):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:68]: Instead of using enumerate and getting by index in the next line, try to use `zip`

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -37,32 +63,75 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+            sum_difference += (value - actual[i]) ** 2
+        mse = sum_difference / len(predicted)
+        return mse
+
+    return None
+
+
+def validate_input(dict1: Any) -> bool:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:76]: You can not add new functions

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -37,32 +63,75 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+    unknown_profile: dict[str, str | dict[str, float]],
+    profile_to_compare: dict[str, str | dict[str, float]]
 ) -> float | None:
     """
     Compares profiles and calculates the distance using symbols
     :param unknown_profile: a dictionary of an unknown profile
     :param profile_to_compare: a dictionary of a profile to compare the unknown profile to
     :return: the distance between the profiles
     """
+    if validate_input(unknown_profile) and validate_input(profile_to_compare):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:95]: early return

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -37,32 +63,75 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+    profile_1: dict[str, str | dict[str, float]],
+    profile_2: dict[str, str | dict[str, float]],
 ) -> str | None:
     """
     Detects the language of an unknown profile
     :param unknown_profile: a dictionary of a profile to determine the language of
     :param profile_1: a dictionary of a known profile
     :param profile_2: a dictionary of a known profile
     :return: a language
+    @rtype: object
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:118]: what is it?

Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:118]: remove!

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -37,32 +63,75 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     """
+    if (
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:120]: early return

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -71,6 +140,12 @@ def load_profile(path_to_file: str) -> dict | None:
     :param path_to_file: a path to the language profile
     :return: a dictionary with at least two keys – name, freq
     """
+    if isinstance(path_to_file, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:143]: early return

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -80,6 +155,19 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
     :return: a dict with a lower-cased loaded profile
     with relative frequencies without unnecessary ngrams
     """
+    if (
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:158]: again

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -80,6 +155,19 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+            isinstance(profile, dict) and "name" in profile
+            and "freq" in profile and "n_words" in profile
+    ):
+        preprocessed_profile = {"name": profile["name"], "freq": {}}
+        unigrams = profile["n_words"][0]
+        for token in profile["freq"]:
+            if len(token.strip()) == 1 and token.isalpha():
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:165]: call strip only once

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -88,20 +176,46 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
     :paths_to_profiles: a list of strings to the profiles
     :return: a list of loaded profiles
     """
+    if isinstance(paths_to_profiles, list):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:179]: again :)

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -88,20 +176,46 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
+        unknown_profile: dict[str, str | dict[str, float]],
+        known_profiles: list
+) -> list | None:
     """
     Detects the language of an unknown profile
     :param unknown_profile: a dictionary of a profile to determine the language of
     :param known_profiles: a list of known profiles
     :return: a sorted list of tuples containing a language and a distance
     """
+    if isinstance(unknown_profile, dict) and isinstance(known_profiles, list):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:198]: and again

File: lab_1_classify_by_unigrams/main.py [hunk 16]
```diff
@@ -88,20 +176,46 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
+        for path in paths_to_profiles:
+            preprocess_profiles_list.append(preprocess_profile(load_profile(path)))
+        return preprocess_profiles_list
+
+    return None
 
 
-def detect_language_advanced(unknown_profile: dict[str, str | dict[str, float]],
-                             known_profiles: list) -> list | None:
+def detect_language_advanced(
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:188]: why did you change it?

File: lab_1_classify_by_unigrams/target_score.txt [hunk 17]
```diff
@@ -1 +1 @@
-0
+8
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/target_score.txt:1]: What about 10?

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:188]: Because code style check didn't like the length of the line)

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: .gitignore [hunk 18]
```diff
@@ -146,3 +146,6 @@ build
 
 # autogenerated stubs
 out
+
+# seminar files
+seminars/
```
Comment (artyomtugaryov) [inline .gitignore:151]: no, revert!

File: lab_1_classify_by_unigrams/main.py [hunk 19]
```diff
@@ -11,6 +12,12 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if not isinstance(text, str):
+        return None
+
+    tokens = [symbol.lower() for symbol in text if symbol.isalpha()]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:18]: return as soon as posible, do not create a variable

File: lab_1_classify_by_unigrams/main.py [hunk 20]
```diff
@@ -19,6 +26,15 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list):
+        return None
+
+    if not all(isinstance(token, str) for token in tokens):
+        return None
+
+    frequency = {token: tokens.count(token) / len(tokens) for token in tokens}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:35]: again

File: lab_1_classify_by_unigrams/main.py [hunk 21]
```diff
@@ -28,6 +44,12 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     :param text: a text
     :return: a dictionary with two keys – name, freq
     """
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+
+    language_profile = {'name': language, 'freq': calculate_frequencies(tokenize(text))}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:50]: the same again

File: lab_1_classify_by_unigrams/main.py [hunk 22]
```diff
@@ -37,24 +59,50 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+    if not (
+            isinstance(predicted, list) and isinstance(actual, list)
+            and len(predicted) == len(actual)
+    ):
+        return None
+
+    sum_difference = 0
+    for freq_value in zip(predicted, actual):
+        sum_difference += (freq_value[0] - freq_value[1]) ** 2
+    mse = sum_difference / len(predicted)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:71]: again:)

File: lab_1_classify_by_unigrams/main.py [hunk 23]
```diff
@@ -37,24 +59,50 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+
+    return mse
 
 
 def compare_profiles(
-        unknown_profile: dict[str, str | dict[str, float]],
-        profile_to_compare: dict[str, str | dict[str, float]]
+    unknown_profile: dict[str, str | dict[str, float]],
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:77]: revert this changes

File: lab_1_classify_by_unigrams/main.py [hunk 24]
```diff
@@ -37,24 +59,50 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+        lang2_freq.append(profile_to_compare['freq'].get(token, 0))
+
+    return calculate_mse(lang1_freq, lang2_freq)
 
 
 def detect_language(
-        unknown_profile: dict[str, str | dict[str, float]],
-        profile_1: dict[str, str | dict[str, float]],
-        profile_2: dict[str, str | dict[str, float]],
+    unknown_profile: dict[str, str | dict[str, float]],
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:103]: and this

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:188]: It is strange. What exactly style checker? On CI?

File: seminars/practice_2_string.py [hunk 25]
```diff
@@ -53,6 +53,9 @@ def multiply_string(input_string: str, how_many: int) -> str:
     display the given string the number of times given in the `how_many`.
     """
     # student realisation goes here
+    format_string = input_string * how_many
```
Comment (artyomtugaryov) [inline seminars/practice_2_string.py:56]: Do not commit seminars. Revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:188]: Yes, on CI

File: lab_1_classify_by_unigrams/target_score.txt (hunk 17 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/target_score.txt]: Changed to 10

File: .gitignore (hunk 18 above)
Comment (velgarita) [reply .gitignore:151]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 19 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:18]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 20 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:35]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 21 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:50]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 22 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:71]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 23 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:77]: It's code style check again

File: lab_1_classify_by_unigrams/main.py (hunk 24 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py:103]: Code style check

File: seminars/practice_2_string.py (hunk 25 above)
Comment (velgarita) [reply seminars/practice_2_string.py]: Done

File: .gitignore [hunk 26]
```diff
@@ -146,3 +146,4 @@ build
 
 # autogenerated stubs
 out
+
```
Comment (artyomtugaryov) [inline .gitignore:149]: revert

File: lab_1_classify_by_unigrams/main.py (hunk 24 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:100]: Hm... Are you syre?

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:196]: Please, provide a proof - link to failed CI

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good, but there are strange changes

File: .gitignore (hunk 26 above)
Comment (velgarita) [reply .gitignore]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: tg_image_1220455557

Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: In my code lines 99 and 197 are two functions: detect_language and detect_language_advanced. That's two lines that I changed because of code style errors.

File: lab_1_classify_by_unigrams/main.py (hunk 24 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: I pined a proof in a comment above.

File: .gitignore [hunk 27]
```diff
@@ -145,4 +145,4 @@ dictionary.dic
 build
 
 # autogenerated stubs
-out
+out
```
Comment (artyomtugaryov) [inline .gitignore:148]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 28]
```diff
@@ -49,12 +80,26 @@ def compare_profiles(
...
+        lang2_freq.append(profile_to_compare['freq'].get(token, 0))
+
+    return calculate_mse(lang1_freq, lang2_freq)
 
 
 def detect_language(
         unknown_profile: dict[str, str | dict[str, float]],
         profile_1: dict[str, str | dict[str, float]],
-        profile_2: dict[str, str | dict[str, float]],
+        profile_2: dict[str, str | dict[str, float]]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:102]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 29]
```diff
@@ -63,6 +108,32 @@ def detect_language(
...
+            isinstance(unknown_profile, dict)
+            and isinstance(profile_1, dict) and isinstance(profile_2, dict)
+    ):
+        return None
+
+    name_1, name_2 = profile_1['name'], profile_2['name']
+    metric_1 = compare_profiles(unknown_profile, profile_1)
+    metric_2 = compare_profiles(unknown_profile, profile_2)
+
+    if (
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:121]: please reverse the condition and return None in `if`

File: lab_1_classify_by_unigrams/main.py [hunk 30]
```diff
@@ -71,6 +142,13 @@ def load_profile(path_to_file: str) -> dict | None:
     :param path_to_file: a path to the language profile
     :return: a dictionary with at least two keys – name, freq
     """
+    if not isinstance(path_to_file, str):
+        return None
+
+    with open(path_to_file, "r", encoding="utf-8") as file:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:148]: file is a bad name -it reserved

File: lab_1_classify_by_unigrams/main.py [hunk 31]
```diff
@@ -80,6 +158,23 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+            and 'freq' in profile and 'n_words' in profile
+    ):
+        return None
+
+    preprocessed_profile = {'name': profile['name'], 'freq': {}}
+    unigrams = profile['n_words'][0]
+    for token in profile['freq']:
+        if token.lower() in preprocessed_profile['freq']:
+            preprocessed_profile['freq'][token.lower()] += \
+                profile['freq'][token] / profile['n_words'][0]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:172]: It should be more readable if you save `profile['freq'][token]`, `token.lower()` to a variables

Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:172]: you create `unigrams` and do not use it here. Why?

File: lab_1_classify_by_unigrams/main.py [hunk 32]
```diff
@@ -88,20 +183,60 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
+
+    preprocess_profiles_list = []
+    for path in paths_to_profiles:
+        profile = load_profile(path)
+        if isinstance(profile, dict):
+            preprocessed_profile = preprocess_profile(profile)
+            if isinstance(preprocessed_profile, dict):
+                preprocess_profiles_list.append(preprocessed_profile)
 
+    if isinstance(preprocess_profiles_list, list):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:199]: looks strange... I guess you should revork conditions in the loop and return None in bad cases

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:205]: Yes

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good. Small changes and you finish!

File: .gitignore (hunk 27 above)
Comment (velgarita) [reply .gitignore]: done

File: lab_1_classify_by_unigrams/main.py (hunk 28 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 29 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 30 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: changed

File: lab_1_classify_by_unigrams/main.py (hunk 32 above)
Comment (velgarita) [reply lab_1_classify_by_unigrams/main.py]: Done
